audb/core/publish.py

```diff
@@ -122,7 +122,7 @@ def _put_media(
     def job(archive):
         if archive in map_media_to_files:
             for file in map_media_to_files[archive]:
-                deps.data[file][define.DependField.VERSION] = version
+                deps.set_value(file, define.DependField.VERSION, version)
```

Comment:
This is the only other place where we use `deps.set_value()`, I haven't checked if we can replace it by something different, but I think I would not make the `set_value()` method public as it seems confusing to me that we also have `add_media()` and `add_meta()`. The other question is if a user should be able at all to mess around with the dependencies?

Reply:
Tricky question. At the moment I don't really see a use-case for a user. On the other hand, in the beginning I thought that the whole `Dependencies` class is not of interest to the user. But since we had the `get_dependency()` function in old `audb`, it made sense to return it in that place. Yet, I don't think that the "normal" user will actually ever directly work with the `Dependencies` class. On other hand, an advanced user who knows what she's doing might find it useful to be able to modify it... probably not too important. But then we should also hide `add_media()` and `add_meta()` of course.

Reply:
It seems I still do not understand what is going on here. I had a look and for me this line could be simply removed:

For every `archive` in `media` you make sure that it is also in `map_media_to_files`. If yes you set the version in the dependency table for all files to the current version.
But if I look a few lines up:
```python

map_media_to_files[deps.archive(file)].append(file)
if deps.version(file) == version:
    media.add(deps.archive(file))
```
which means if an archive is inside `map_media_to_files` and its version is already `version` only then it is added to `media`, which is then used as the starting point of the loop I discuss above. Which means the version should already match `version`.

But when i remove the `deps.set_value(file, define.DependField.VERSION, version)` line, the test fails.

Reply:
I would propose for a start to add a few more comments to the code what is going on.

Reply:
> which means if an archive is inside `map_media_to_files` and its version is already `version` only then it is added to `media`, which is then used as the starting point of the loop I discuss above. Which means the version should already match `version`.

But it does not for those media that is not altered in this version, but in the same archive as an altered media file.

> I would propose for a start to add a few more comments to the code what is going on.

Fair enough, I will add some more comments :)

Reply:
Ok, I had a look and I think I know what makes the code so hard to understand. Originally we had a single function that would first figure out which files are new/altered and then uploads them. And some point we split it up into to functions `_find_[tables,media]()` and `_put_[tables,media]()`. For the table case this was easy because we have a one table - one archive mapping and the result of the function is a list with all tables we need to upload which we then pass to `_put_media()`. But for `media` it's more complicated because several media files may be in the same archive, which means we have to upload an archive if one media file is altered and also if one media file is removed (and in that case we also have to update the version of all media files in the archive, because they are all uploaded again). So `file_media()` currently only returns a list of archives that contain a media that has been removed and it sets the version for files that are new or have been altered to the version of the new release. Then in `_put_media()` as a first step we figure out the archives that contain new or altered media files by checking if their version equals the version of the new release. But yeah, this should actually be done in `_find_media()` already. But now I remember that I wanted to avoid create this `map_media_to_files` dictionary twice. But now I see that this needs to be changed, it's too complicated for other developers. It will move some parts of `_put_media()` to `_find_media()` and probably pass that `map_media_to_files` between the functions, but that's not a big deal of course. But I suggest I do that after we have merged this and the other PRs to avoid conflicts.

Reply:
I created https://github.com/audeering/audb/issues/53

Reply:
OK, but even with cleaning the code up this means we cannot avoid `deps.set_value()` because we have to update the dependencies.
Could we maybe instead call:
```python
deps.add_media(db_root, file, archive, checksum, version)
```
as we do at another place to update a media file?

Reply:
Yes, let's get rid of `set_value()` and always use the according `add_*()` function instead.

Reply:
I was able to replace `set_value()` by `_add_meta()` and `_add_media()`. The later I modified slightly so that the audio metadata information is not collected when we want to change only the version as it is happening here.